### PR TITLE
Exercise 3

### PR DESCRIPTION
**Pull request by Tasnima123**

### CODE REVIEW

File: .env.test
Comment (JackMeadDev) [inline .env.test:7]: These look like real credentials, but from a quick test it looks like they probably aren't - is that right? In future I'd recommend using dummy variables e.g. `test-api-key` or something, so that it's really obvious to everybody that they aren't real credentials

File: README.md
Comment (JackMeadDev) [inline README.md]: We should update the Readme file to let new developers know how they can run the tests

File: geckodriver.log
Comment (JackMeadDev) [inline geckodriver.log]: This file could be included in the .gitignore, as it's unlikely to be of value to other developers on the team

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: This is a nice example of a view model! We should probably separate it out into its own file to make it more readable, and similarly with the `Item` class

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: We should really pull these out into environment variables too, so that if a new trello account/board is being used we don't have to search the code to update these list ids

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: Similarly here, you store the `TRELLO_BOARD_ID` environment variable already, so we should use that version of the variable rather than an explicit string, so that e.g. tests can use an alternative board easily

File: todo_app/tests/test_IntUnit.py
Comment (JackMeadDev) [inline todo_app/tests/test_IntUnit.py]: It would be good to move this loading of dotenv into the `test_app` method, as that's strictly where it's required, and this can be confusing (e.g, in your e2e test, if run from the command line alone it doesn't currently load the .env.test variables in)

File: todo_app/tests_e2e/test_e2e.py
Comment (JackMeadDev) [inline todo_app/tests_e2e/test_e2e.py]: Nice - we should add some extra checks here; can you get it to add a new item and check that the new item now exists?

File: todo_app/app.py
Comment (JackMeadDev) [inline todo_app/app.py]: Nice, it's good to be able to create/delete boards for our tests, but currently I can't see anywhere this is used. It might be clearer to include this method explicitly in our `tests_e2e` folder too, so that its clear what the purpose/intention is

File: todo_app/tests/test_IntUnit.py
Comment (JackMeadDev) [inline todo_app/tests/test_IntUnit.py]: This test feels a little confused, currently this hits our index page then checks that the `get_requests` method defined just above returns an item with a title

It might be worth taking a little look at [this article](https://realpython.com/python-mock-library/) under "Patch as a Decorator" to see how we configure the mock; we'll then want to check that the mocked item ends up on the page we loaded (i.e. is in the response), rather than testing the mocked function directly.

**Review by JackMeadDev — CHANGES_REQUESTED** [review]
Nice work, we have some functioning tests!

I've made a few comments in the pull request, and I think there are still a few changes we'll want to update before we merge this. In particular:
* We should make it clear in the ReadMe how to set up & run the tests
* We should extend the end-to-end test to interact with the page
  * We might need to pull out some of the trello ids that are specified in the code to help with this
  * And we'll need to take advantage of the create/delete board functions you specified
* We should change the integration tests assertion to check the right items have been picked up
* We should add a unit test, maybe testing the ViewModel properties are working as expected?

There's a lot going on in this module, so please do shout out if you need any help with the above, and particularly feel free to arrange a 1-1 call if useful!

File: .env.template
Comment (hugh-emerson) [inline .env.template]: This file shouldn't be deleted. It's useful to have a starting point for creating a .env file. And in fact the README still entions it - `cp .env.template .env`

File: README.md
Comment (hugh-emerson) [inline README.md]: Needs to be `poetry run pytest` if your shell isn't in the poetry env already

File: geckodriver.log
Comment (hugh-emerson) [reply geckodriver.log]: This has not been done yet. geckodriver should also be in the .gitignore.

Also after adding them to the gitignore, you will need to remove them from the repo since they're already tracked.

File: pyproject.toml
Comment (hugh-emerson) [inline pyproject.toml:15]: Why are these here?

I expected to see pytest and selenium

File: todo_app/*.geckodriver.log
Comment (hugh-emerson) [inline todo_app/*.geckodriver.log]: Did you accidentally create this file? It actually prevents me from checking out this branch because of the strange filename

File: todo_app/app.py
Comment (hugh-emerson) [inline todo_app/app.py]: This doesn't use share anything else in the file and isn't really part of the app so it belongs in the test file rather than here.

File: todo_app/tests_e2e/test_e2e.py
Comment (hugh-emerson) [inline todo_app/tests_e2e/test_e2e.py]: Slightly confusing to update os.environ["TRELLO_BOARD_ID"] but update other env vars (e.g. `done_status`) inside the create_trello_board function

File: todo_app/tests/test_IntUnit.py
Comment (hugh-emerson) [inline todo_app/tests/test_IntUnit.py]: These three tests aren't doing anything meaningful. They are each creating a list with a single dictionary in it and then asserting the list has length == 1.
They are not using any of your application code.

You should have a unit test which creates a `view_model = ViewModel(...)` with some item data and then assert that its properties are correct. E.g. `assert view_model.statusToDo == ...`

File: README.md
Comment (hugh-emerson) [inline README.md]: You should mention the requirement of having Firefox installed and a matching version of geckodriver

**Review by hugh-emerson — CHANGES_REQUESTED** [review]
Almost there but a few blocking issues:
- You need to add pytest and selenium to pyproject.toml (install them with `poetry add pytest selenium --dev`)
- Your unit tests aren't testing any app code
- Your geckodriver.log files need to be removed (and actually the one with the asterisk in the name prevents my git from checking out this branch)

Otherwise it's looking good and I've just left some minor comments.

File: .env.template
Comment (JackMeadDev) [inline .env.template]: Minor, but now you have environment variables for your list ids it would be nice to include them in the `.env.template` file

File: README.md
Comment (JackMeadDev) [inline README.md]: Worth noting that you've moved these folders, so you'll now need to run e.g. `pytest todo_app/tests`

File: README.md
Comment (JackMeadDev) [inline README.md]: I suspect this is meant to be `if *not* in poetry env`, and also worth noting that the command requires an extra `run` i.e. `poetry run pytest <folder>`

File: todo_app/tests/test_IntUnit.py
Comment (JackMeadDev) [inline todo_app/tests/test_IntUnit.py]: These are now testing your app code, but it's a slightly odd use of the properties: we should avoid ever calling variables prefixed with `_` outside of the relevant class, that prefix suggests that outside users shouldn't need to know or care about that
We can replace that by just verifying the properties directly:
```
       assert len(view_model.statusToDo) == 1
       assert len(view_model.statusDoing) == 1
       assert len(view_model.show_all_done_items) == 1
```

**Review by JackMeadDev — APPROVED** [review]
Thanks for fixing up those issues, I've left a few comments on things that would be worth adjusting (particularly making sure the Readme commands still match your codebase) but I'm happy for you to merge this when you're ready.

File: README.md
Comment (Tasnima123) [reply README.md]: moved these to outside the todo_app folder so running pytest tests works.

**Pull request merged**
